src/contract.rs

```diff
@@ -242,11 +248,98 @@ pub fn query<S: Storage, A: Api, Q: Querier>(deps: &Extern<S, A, Q>, msg: QueryM
         QueryMsg::ContractStatus {} => query_contract_status(&deps.storage),
         QueryMsg::ExchangeRate {} => query_exchange_rate(&deps.storage),
         QueryMsg::Minters { .. } => query_minters(deps),
-        _ => authenticated_queries(deps, msg),
+        QueryMsg::WithPermit { permit, query } => query_with_permit(deps, permit, query),
+        _ => viewing_keys_queries(deps, msg),
     }
 }
 
-pub fn authenticated_queries<S: Storage, A: Api, Q: Querier>(
+fn query_with_permit<S: Storage, A: Api, Q: Querier>(
+    deps: &Extern<S, A, Q>,
+    permit: Permit,
+    query: QueryWithPermit,
+) -> Result<Binary, StdError> {
+    if permit.signed.msgs.len() != 1 {
+        return Err(StdError::generic_err(format!(
+            "Must sign exactly 1 permit message, got: {:?}.",
+            permit.signed.msgs.len()
+        )));
+    }
+
+    if permit.signed.msgs[0].r#type != "query_permit" {
+        return Err(StdError::generic_err(format!(
+            "Type must be 'query_permit', got: {:?}.",
+            permit.signed.msgs[0].r#type
+        )));
+    }
+
+    let permit_content = &permit.signed.msgs[0].value;
```

Comment:
So, (in my opinion) instead of all this verification, `WithPermit::signed` should be of type `PermitContent` (and maybe renamed from `signed` to something else).
Then create the `SignedPermit` instance based on that with something like `SignedPermit::from_content` which fills in those default blank values. That way the query is easier to write (imagine if we want to query this manually from the CLI?) and faster to parse as well (minor but still)

Reply:
This is auto-generated from the signing process on the client, I'm not changing the names here, it'll just complicate client code.

Reply:
I don't think you understood, let's discuss tomorrow

Reply:
#23